texar/modules/pretrained/xlnet_model_utils.py

```diff
@@ -0,0 +1,551 @@
+            tf.get_variable_scope().set_initializer(
+                layers.get_initializer(self._hparams.initializer))
+            l1_hparams = {
+                "type": "Dense",
+                "kwargs": {
+                    "units": ffn_inner_dim,
+                    "activation": activation
+                }
+            }
+            self.linear1 = layers.get_layer(hparams=l1_hparams)
+            dropout_hparams = {
+                "type": "Dropout",
+                "kwargs": {
+                    "rate": dropout
+                }
+            }
+            self.dropout = layers.get_layer(hparams=dropout_hparams)
+            l2_hparams = {
+                "type": "Dense",
+                "kwargs": {
+                    "units": hidden_dim
+                }
+            }
+            self.linear2 = layers.get_layer(hparams=l2_hparams)
+
+    @staticmethod
+    def default_hparams():
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        .. code-block:: python
+
+            {
+                "hidden_dim": 768,
+                "ffn_inner_dim": 3072,
+                "dropout": 0.1,
+                "activation": 'gelu'
+            }
+
+        Here
+
+        "hidden_dim": int
+            Dimension of the layer fed as input to feed forward network
+
+        "ffn_inner_dim": int
+            Inner dimension of the feed forward layer
+
+        "dropout": float
+            Dropout rate for layers
+
+        "activation": str or callable
+            Activation function applied to the output of the PositionWise FF.
+            See :func:`~texar.core.get_activation_fn` for more details.
+        """
+        return {
+            "name": "ff",
+            "initializer": None,
+            "hidden_dim": 768,
+            "ffn_inner_dim": 3072,
+            "dropout": 0.1,
+            "activation": 'gelu',
+        }
+
+    def _build(self, input, mode=None):
+        r"""Compute feed forward for the input.
+
+        Args:
+            input: Input tensor of size `(max_time, batch_size, hidden_dim)`
+            mode (optional): A tensor taking value in
+                :tf_main:`tf.estimator.ModeKeys <estimator/ModeKeys>`, including
+                `TRAIN`, `EVAL`, and `PREDICT`. If `None`, dropout is
+                controlled by :func:`texar.global_mode`.
+
+        :returns: A tensor output of the position wise feed forward network
+        """
+        is_training = is_train_mode(mode)
+        output = self.linear1(input)
+        output = self.dropout(output, training=is_training)
+        output = self.linear2(output)
+        output = self.dropout(output, training=is_training)
+
+        # residual + layer norm
+        output = tf.contrib.layers.layer_norm(
+            input + output, begin_norm_axis=-1, scope=self.variable_scope,
+            reuse=tf.AUTO_REUSE)
+
+        return output
+
+
+class PositionalEmbedding(ModuleBase):
+    r"""Sinosoidal Positional Embedding."""
```

Comment:
This embedding must be moved to `embedder` and if possible, mered with the current Sinosoidal position embedder, pls open an issue to record this TODO

Reply:
Done. Issue https://github.com/asyml/texar/issues/195

Reply:
Added a TODO. Issue https://github.com/asyml/texar/issues/195